File: code/client/munkilib/updatecheck/licensing.py
# encoding: utf-8
#
# Copyright 2009-2023 Greg Neagle.
#
# Licensed under the Apache License, Version 2.0 (the 'License');
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#      https://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an 'AS IS' BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
"""
updatecheck.licensing

Created by Greg Neagle on 2017-01-01.

"""
from __future__ import absolute_import, print_function

try:
    from urllib import quote_plus
except ImportError:
    # Python 3
    from urllib.parse import quote_plus

from .. import display
from .. import fetch
from .. import prefs
from .. import FoundationPlist


def update_available_license_seats(installinfo):
    '''Records # of available seats for each optional install'''

    license_info_url = prefs.pref('LicenseInfoURL')
    if not license_info_url:
        # nothing to do!
        return
    if not installinfo.get('optional_installs'):
        # nothing to do!
        return

    license_info = {}
    items_to_check = [item['name']
                      for item in installinfo['optional_installs']
                      if item.get('licensed_seat_info_available')
                      and not item['installed']]

    # complicated logic here to 'batch' process our GET requests but
    # keep them under 256 characters each
    start_index = 0
    # Use ampersand when the license_info_url contains a ?
    q_char = "?"
    if "?" in license_info_url:
        q_char = "&"
    while start_index < len(items_to_check):
        end_index = len(items_to_check)
        while True:
            query_items = ['name=' + quote_plus(item)
                           for item in items_to_check[start_index:end_index]]
            url = license_info_url + q_char + '&'.join(query_items)
            if len(url) < 256:
                break
            # drop an item and see if we're under 256 characters
            end_index = end_index - 1

        display.display_debug1('Fetching licensed seat data from %s', url)
        try:
            license_data = fetch.getDataFromURL(url)
            display.display_debug1('Got: %s', license_data)
            license_dict = FoundationPlist.readPlistFromString(
                license_data.encode("UTF-8"))
        except fetch.Error as err:
            # problem fetching from URL
            display.display_error('Error from %s: %s', url, err)
        except FoundationPlist.FoundationPlistException:
            # no data or bad data from URL
            display.display_error(
                'Bad license data from %s: %s', url, license_data)
        else:
            # merge data from license_dict into license_info
            license_info.update(license_dict)
        start_index = end_index

    # use license_info to update our remaining seats
    for item in installinfo['optional_installs']:
        if item['name'] in items_to_check:
            display.display_debug2(
                'Looking for license info for %s', item['name'])
            # record available seats for this item
            seats_available = False
            seat_info = license_info.get(item['name'], 0)
            try:
                seats_available = int(seat_info) > 0
                display.display_debug1(
                    'Recording available seats for %s: %s',
                    item['name'], seats_available)
            except ValueError:
                display.display_warning(
                    'Bad license data for %s: %s', item['name'], seat_info)

            item['licensed_seats_available'] = seats_available


if __name__ == '__main__':
    print('This is a library of support tools for the Munki Suite.')

File: code/cli/munki/shared/updatecheck/licensing.swift
//
//  licensing.swift
//  munki
//
//  Created by Greg Neagle on 8/20/24.
//  Copyright 2024-2026 The Munki Project. All rights reserved.
//
//  Licensed under the Apache License, Version 2.0 (the "License");
//  you may not use this file except in compliance with the License.
//  You may obtain a copy of the License at
//
//       https://www.apache.org/licenses/LICENSE-2.0
//
//  Unless required by applicable law or agreed to in writing, software
//  distributed under the License is distributed on an "AS IS" BASIS,
//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
//  See the License for the specific language governing permissions and
//  limitations under the License.

import Foundation

/// Records # of available seats for each optional install
/// returns an updated list of optional_installs with license info
func updateAvailableLicenseSeats(_ optionalInstalls: [PlistDict]) -> [PlistDict] {
    guard let licenseInfoURL = pref("LicenseInfoURL") as? String else {
        // nothing to do
        return optionalInstalls
    }

    let display = DisplayAndLog.main
    var mutableOptionalInstalls = optionalInstalls

    var licenseInfo = PlistDict()
    let itemsToCheck: [String]
    itemsToCheck = mutableOptionalInstalls.filter {
        ($0["licensed_seat_info_available"] as? Bool ?? false) &&
            !($0["installed"] as? Bool ?? false)
    }.map {
        $0["name"] as? String ?? ""
    }.filter {
        !$0.isEmpty
    }
    let queryItems = itemsToCheck.map {
        "name=\(($0 as NSString).addingPercentEncoding(withAllowedCharacters: .urlQueryAllowed) ?? "<invalidname>")"
    }

    // complicated logic here to 'batch' process our GET requests but
    // keep them under 256 characters each
    var startIndex = 0
    // Use ampersand when the license_info_url contains a ?
    let qChar = if licenseInfoURL.contains("?") {
        "&"
    } else {
        "?"
    }
    while startIndex < itemsToCheck.count {
        var endIndex = itemsToCheck.count
        var url = ""
        while true {
            url = licenseInfoURL + qChar + queryItems[startIndex ..< endIndex].joined(separator: "&")
            if url.count < 256 {
                break
            }
            // too long; drop an item and see if we're under 256 characters
            endIndex -= 1
        }
        display.debug1("Fetching licensed seat data from \(url)")
        do {
            if let licenseData = try getDataFromURL(url) {
                display.debug1("Got: \(String(data: licenseData, encoding: .utf8) ?? "")")
                if let licenseDict = try readPlist(fromData: licenseData) as? PlistDict {
                    licenseInfo.merge(licenseDict) { _, second in second }
                }
            }
        } catch {
            display.error("Error getting license data: \(error.localizedDescription)")
        }
        // next loop, start where we ended
        startIndex = endIndex
    }

    // use licenseInfo to update remaining seats
    for (index, item) in mutableOptionalInstalls.enumerated() {
        guard let itemName = item["name"] as? String else {
            continue
        }
        if itemsToCheck.contains(itemName) {
            display.debug2("Looking for license info for \(itemName)")
            var seatsAvailable = false
            let seatInfo = licenseInfo[itemName] as? Int ?? 0
            display.debug1("\(seatInfo) seats available for \(itemName)")
            if seatInfo > 0 {
                seatsAvailable = true
            }
            mutableOptionalInstalls[index]["licensed_seats_available"] = seatsAvailable
        }
    }
    return mutableOptionalInstalls
}
